### PR TITLE
Add helm chart

### PR DESCRIPTION
**Pull request by devec0**
Hello,

Thank you for writing k8s-ecr-login-renew, and sharing it! It definitely saved me some heartache today.

Noticing #23 - and having decent experience with writing helm charts - and needing to deploy this via Helm myself, I thought I'd write and contribute a chart to this project. As noted in #23 - people who are maintaining complex Kubernetes deployments are quite commonly using Helm due to the templating and secret-handling utility is has, as well as companion tools such as `helmfile` to help keep multi-environment deployments as DRY as possible. This was my use case, so I hope this chart will help others.

I've tested this chart myself and it is working well.

I've tried to keep it as minimal and secure as possible following the good example in the `examples` folder. I have not added the ability to specify multiple repositories to the chart, but that would be quite easy to add later if someone needed it in a non-breaking way.

Let me know if any questions or if you have any changes you'd like to see.

### CODE REVIEW

**Comment by nabsul** [comment]
Thanks for the contribution @devec0 ! As someone who doesn't use it at all, I'm hoping looking forward to using your PR to learn more about Helm. 

I'll try to review this ASAP, but please be patient. If you don't hear back from me in a couple of weeks, please feel free to ping me with a reminder.

Thanks again!

**Comment by xavidop** [comment]
thanks for this contribution @devec0 ! this is so much needed!

File: chart/templates/cron.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+              - name: AWS_REGION
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: AWS_REGION
+              - name: DOCKER_SECRET_NAME
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: DOCKER_SECRET_NAME
```
Comment (xavidop) [inline chart/templates/cron.yaml:45]: here you have missed:
```
              - name: TARGET_NAMESPACE
                value: {{ .Release.Namespace }}
```
if not it will apply on the default namespace

**Review by xavidop — CHANGES_REQUESTED** [review]

**Comment by armenr** [comment]
I was about to add the same thing today, @xavidop - 

The templates are dangerous without the explicit `.Release.Namespace`

Have a look at mine. Thank you BOTH for YOUR hard work!

https://github.com/armenr/helm-k8s-ecr-login-renew

File: chart/templates/cron.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+apiVersion: batch/v1
+kind: CronJob
+metadata:
+  name: {{ include "k8s-ecr-login-renew.cronJob" . }}
+  labels:
+{{ include "k8s-ecr-login-renew.labels" . | indent 4 }}
+spec:
+  schedule: "* */6 * * *"
```
Comment (PawelLipski) [inline chart/templates/cron.yaml:8]: One thing! I'd suggest adding a `pre-install` Helm hook (a Pod)... b/c otherwise now the credentails won't be filled up until the first run of the cron job :/

File: chart/templates/cron.yaml (hunk 1 above)
Comment (devec0) [reply chart/templates/cron.yaml:45]: Ah, yep, I hit the same issue and ended up adding this to my local copy. I'll push those changes.

File: chart/templates/cron.yaml (hunk 2 above)
Comment (devec0) [reply chart/templates/cron.yaml:8]: Thanks, this is a good idea. I'll push this change in a sec.

**Comment by devec0** [comment]
I've pushed some updates including -

* Passing the deployment namespace to the job by default
* Allowing override of the namespace from helm configuration explicitly
* Adding a pre-install hook that creates the job for immediate renewal on install

Thanks @PawelLipski and @xavidop for the helpful review and kind words.

**Comment by devec0** [comment]
Additionally, if anyone has changes they would like to make, and would like push access to the forked version I have attached to this PR, either open a PR on my repo and I'll merge and push it - or just please let me know, I will grant push access if needed.

**Comment by nabsul** [comment]
I did start learning Helm yesterday finally :-). Hopefully in the next week or two I'll be in a position to review and merge this.

**Comment by devec0** [comment]
Please take your time @nabsul - helm takes some time to get used to, but it is worth it, and thanks for considering this PR 🙂

File: chart/templates/cron.yaml [hunk 3]
```diff
@@ -43,3 +43,20 @@ spec:
                   secretKeyRef:
                     name: {{ include "k8s-ecr-login-renew.secret" . }}
                     key: DOCKER_SECRET_NAME
+{{- if .Values.overrideNamespace }}
+              - name: TARGET_NAMESPACE
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
```
Comment (PawelLipski) [inline chart/templates/cron.yaml:50]: Hmmm I don't think it makes sense to read namespace name from a secret 🤔 as in, it's much better retrieved from Helm values directly

Comment (PawelLipski) [reply chart/templates/cron.yaml:50]: Actually, in my similar chart, I'm using a single instance of renewer for multiple namespaces:

```
                - name: TARGET_NAMESPACE
                  # Despite a singular name of the env var itself,
                  # multiple namespaces can be provided - see https://github.com/nabsul/k8s-ecr-login-renew#multiple-namespace-support
                  value: {{ join "," .Values.targetNamespaces }}
```

File: chart/templates/cron.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+spec:
+  schedule: "* */6 * * *"
+  successfulJobsHistoryLimit: 3
+  failedJobsHistoryLimit: 5
+  jobTemplate:
+    spec:
+      template:
+        metadata:
+          labels:
+{{ include "k8s-ecr-login-renew.podLabels" . | indent 12 }}
```
Comment (PawelLipski) [inline chart/templates/cron.yaml:16]: Pls keep the `{{` actions indented more in line with the surrounding YAML... e.g. here I'd start the `{{` on the level one deeper than the `labels:` above

File: chart/templates/cron.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: AWS_REGION
+              - name: DOCKER_SECRET_NAME
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: DOCKER_SECRET_NAME
+{{- if .Values.overrideNamespace }}
```
Comment (PawelLipski) [inline chart/templates/cron.yaml:46]: Similarly for the `if`s, pls indent for better readability

File: chart/templates/cron.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                    key: TARGET_NAMESPACE
+{{- else }}
+              - name: TARGET_NAMESPACE
+                value: {{ .Release.Namespace }}
+{{- end }}
+{{- if .Values.ecr.dockerRegistries }}
+              - name: DOCKER_REGISTRIES
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
```
Comment (PawelLipski) [inline chart/templates/cron.yaml:60]: Yeah... I don't think any of these values, other than `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` needs to be provided via `secretKeyRef`. For the reference, here's how it looks like in my chart currently:

```yaml
      env:
        - name: AWS_ACCESS_KEY_ID
          valueFrom:
            secretKeyRef:
              name: aws-ecr-credentials
              key: accessKeyId
        - name: AWS_REGION
          value: eu-west-1
        - name: AWS_SECRET_ACCESS_KEY
          valueFrom:
            secretKeyRef:
              name: aws-ecr-credentials
              key: secretAccessKey
        - name: DOCKER_REGISTRIES
          value: "{{ .Values.ecrRegistry }}"
        - name: DOCKER_SECRET_NAME
          value: ecr-docker-credentials
        - name: TARGET_NAMESPACE
          value: "{{ join "," .Values.targetNamespaces }}"
```

File: chart/templates/serviceaccount.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ include "k8s-ecr-login-renew.roleBinding" . }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ include "k8s-ecr-login-renew.role" . }}
+subjects:
+  - kind: ServiceAccount
```
Comment (PawelLipski) [inline chart/templates/serviceaccount.yaml:24]: Hmmm what about the target namespace here? 🤔 

Compare:

```yaml
kind: ServiceAccount
apiVersion: v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ .Release.Namespace }}"
  annotations:
    "helm.sh/hook": pre-install
    # We use a smaller weight than for ecr-docker-credentials-renewer pod so that the resources are created before the pod.
    "helm.sh/hook-weight": "10"
    # We can't delete the service account (and other resources defined in this file)
    # after running the hook since it's needed for the CronJob as well.
    "helm.sh/hook-delete-policy": before-hook-creation


{{- range .Values.targetNamespaces }}
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ . }}"
  annotations:
    "helm.sh/hook": pre-install
    "helm.sh/hook-weight": "20"
    "helm.sh/hook-delete-policy": before-hook-creation
rules:
  - apiGroups: [""]
    resources: [secrets]
    verbs: [create, update, get, delete]

---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ . }}"
  annotations:
    "helm.sh/hook": pre-install
    "helm.sh/hook-weight": "30"
    "helm.sh/hook-delete-policy": before-hook-creation
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ecr-docker-credentials-renewer
subjects:
  - kind: ServiceAccount
    name: ecr-docker-credentials-renewer
    namespace: "{{ $.Release.Namespace }}"
{{- end }}
```

File: chart/templates/serviceaccount.yaml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  name: {{ include "k8s-ecr-login-renew.role" . }}
+rules:
+  - apiGroups: [""]
+    resources: ["secrets"]
+    verbs: ["create", "update", "get", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ include "k8s-ecr-login-renew.roleBinding" . }}
```
Comment (PawelLipski) [inline chart/templates/serviceaccount.yaml:18]: Hmmm what about the target namespace here? 🤔 

Compare:

```yaml
kind: ServiceAccount
apiVersion: v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ .Release.Namespace }}"
  annotations:
    "helm.sh/hook": pre-install
    # We use a smaller weight than for ecr-docker-credentials-renewer pod so that the resources are created before the pod.
    "helm.sh/hook-weight": "10"
    # We can't delete the service account (and other resources defined in this file)
    # after running the hook since it's needed for the CronJob as well.
    "helm.sh/hook-delete-policy": before-hook-creation


{{- range .Values.targetNamespaces }}
---
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ . }}"
  annotations:
    "helm.sh/hook": pre-install
    "helm.sh/hook-weight": "20"
    "helm.sh/hook-delete-policy": before-hook-creation
rules:
  - apiGroups: [""]
    resources: [secrets]
    verbs: [create, update, get, delete]

---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: ecr-docker-credentials-renewer
  namespace: "{{ . }}"
  annotations:
    "helm.sh/hook": pre-install
    "helm.sh/hook-weight": "30"
    "helm.sh/hook-delete-policy": before-hook-creation
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: ecr-docker-credentials-renewer
subjects:
  - kind: ServiceAccount
    name: ecr-docker-credentials-renewer
    namespace: "{{ $.Release.Namespace }}"
{{- end }}
```

File: chart/templates/cron.yaml (hunk 3 above)
Comment (armenr) [reply chart/templates/cron.yaml:50]: This is a good idea. 👍🏼

**Comment by PawelLipski** [comment]
Hey there... any updates here? ;)

**Comment by nabsul** [comment]
My update from the last three months:

- [ ] Prepare for arrival of baby (never completely finished, but as much as possible)
- [ ] Prepare for parental leave by wrapping up / handing over work projects
- [ ] Baby arrives
- [ ] Get used to new life with baby
- [ ] Catch up on open source projects (this PR is the first priority)
- [ ] Go back to work (mid-september)

I make no promises, but I really do want to get this done, and I think I'll have the energy (and need the distraction from baby-work) as I adjust to baby life. 🤞 🤞 😄

**Comment by devec0** [comment]
> My update from the last three months:
> 
>     * [x]  Prepare for arrival of baby (never completely finished, but as much as possible)
> 
>     * [x]  Prepare for parental leave by wrapping up / handing over work projects
> 
>     * [x]  Baby arrives
> 
>     * [ ]  Get used to new life with baby
> 
>     * [ ]  Catch up on open source projects (this PR is the first priority)
> 
>     * [ ]  Go back to work (mid-september)
> 
> 
> I make no promises, but I really do want to get this done, and I think I'll have the energy (and need the distraction from baby-work) as I adjust to baby life. 🤞 🤞 😄

Congratulations @nabsul 🎉 🎊 
It looks to me like your priorities are correct 👍🏻

**Comment by PawelLipski** [comment]
Wow congrats indeed! No rush, take a good care of the baby 👶🏻 
Girl or boy btw? ;)

**Comment by nabsul** [comment]
Boy :-) And thank you all for all the kind words ❤️

So many decisions to make with this new child, like: Do I teach him Go or C# first? 😆

File: chart/templates/secret.yaml [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  AWS_ACCESS_KEY_ID: {{ .Values.ecr.accessKeyId }}
+  AWS_SECRET_ACCESS_KEY: {{ .Values.ecr.secretAccessKey }}
+  AWS_REGION: {{ .Values.ecr.region }}
+  DOCKER_SECRET_NAME: {{ .Values.secretName }}
+{{- if .Values.dockerRegistries }}
+  DOCKER_REGISTRIES: {{ .Values.ecr.dockerRegistries }}
+{{- end }}
+{{- if .Values.overrideNamespace }}
+  TARGET_NAMESPACE: {{ .Values.overrideNamespace }}
+{{- end }}
```
Comment (Startouf) [inline chart/templates/secret.yaml:18]: I have a very mitigated opinion regarding the generation of this secret via values and security, and this is why I don't like existing charts. 

If you create a secret via Helm values, this kills all confidentiality, as you will be able to access the secret by reading the released chart user-provided values (correct me if I'm wrong).

IMO, the creation of this secret should be left at the user's convenience, and NOT through helm value. Which is why I suggest to make the secret name and key configurable by the user
(or have the values of the generated secret be themselves secretKeyRef for AWS AK/SK), and just add a notice in the readme with a sample secret.yaml that the user needs to apply directly with kubectl apply

What do you think ?

File: chart/values.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+secretName: docker-ecr
+# overrideNamespace:
+
+ecr:
+  region: ""
+  accessKeyId: ""
+  secretAccessKey: ""
```
Comment (Startouf) [inline chart/values.yaml:7]: Regarding my previous question and confidentiality of the secret

```suggestion
  secretName: ""
  awsAccessKeyID_key: ""
  awsSecretAccessKey_key
```

File: chart/templates/cron.yaml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+              - name: AWS_ACCESS_KEY_ID
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: AWS_ACCESS_KEY_ID
+              - name: AWS_SECRET_ACCESS_KEY
+                valueFrom:
+                  secretKeyRef:
+                    name: {{ include "k8s-ecr-login-renew.secret" . }}
+                    key: AWS_SECRET_ACCESS_KEY
```
Comment (Startouf) [inline chart/templates/cron.yaml:35]: See my previous comment

Is it possible to make the secret name + keys configurable via the values ? I like to reuse existing AWS credentials secrets across several apps like this and it's impossible if they each have a different name/key.

(don't forget to propagate to the pre-install-job as well)

```suggestion
                    name: {{ include "k8s-ecr-login-renew.secret" . }}
                    key: {{ .Values.ecr.AwsAccessKeyId_key }}
              - name: AWS_SECRET_ACCESS_KEY
                valueFrom:
                  secretKeyRef:
                    name: {{ include "k8s-ecr-login-renew.secret" . }}
                    key: {{ .Values.ecr.AwsSecretAccessKey_key }}
```

File: chart/templates/definitions.tpl [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+{{- define "k8s-ecr-login-renew.job" }}
+{{- default (printf "%s-job" .Release.Name) -}}
+{{- end }}
+
+{{- define "k8s-ecr-login-renew.cronJob" }}
+{{- default (printf "%s-cron" .Release.Name) -}}
+{{- end }}
+
+{{- define "k8s-ecr-login-renew.secret" }}
+{{- default (printf "%s-secret" .Release.Name) -}}
```
Comment (Startouf) [inline chart/templates/definitions.tpl:30]:
```suggestion
{{- .Values.ecr.secretName | default (printf "%s-secret" .Release.Name) -}}
```

**Review by Startouf — CHANGES_REQUESTED** [review]
I suggest not creating the secret via values to preserve confidentiality. A note could be added to the readme explaining how to create the secret

```
# secrets.nocommit.yaml
apiVersion: v1
kind: Secret
metadata:
  name: aws-secrets
  namespace: xxx
type: Opaque
data:
  aws-acces-key-id: xxx=
  aws-secret-access-key: xxx=
  
# kubectl apply -f ./secrets.nocommit.yaml
```

What do you think ?

File: chart/templates/definitions.tpl (hunk 12 above)
Comment (Startouf) [reply chart/templates/definitions.tpl:30]: Actually if you want to follow my approach, we can completely remove this from the helper

File: chart/templates/secret.yaml (hunk 9 above)
Comment (devec0) [reply chart/templates/secret.yaml:18]: For helm3, the values are stored as a secret in k8s, so it's not really an issue for storing the secrets in helm values.

I do agree it would be good to make the secret creation optional, as some people will use other tools (such as terraform) to precreate the secret. If you want to put together a PR for this against my fork, I'll merge it into this PR.

Comment (Startouf) [reply chart/templates/secret.yaml:18]: @devec0 been working on this, I'm not using windows so it would be better is someone else coud run tests https://github.com/devec0/k8s-ecr-login-renew/pull/1

**Comment by devec0** [comment]
> I suggest not creating the secret via values to preserve confidentiality. A note could be added to the readme explaining how to create the secret
> 
> ```
> # secrets.nocommit.yaml
> apiVersion: v1
> kind: Secret
> metadata:
>   name: aws-secrets
>   namespace: xxx
> type: Opaque
> data:
>   aws-acces-key-id: xxx=
>   aws-secret-access-key: xxx=
>   
> # kubectl apply -f ./secrets.nocommit.yaml
> ```
> 
> What do you think ?

Thank you for your fantastic improvements and suggestions @Startouf - I've reviewed them and merged them into this PR.

**Comment by nabsul** [comment]
> I suggest not creating the secret via values to preserve confidentiality. A note could be added to the readme explaining how to create the secret
> 
> ```
> # secrets.nocommit.yaml
> apiVersion: v1
> kind: Secret
> metadata:
>   name: aws-secrets
>   namespace: xxx
> type: Opaque
> data:
>   aws-acces-key-id: xxx=
>   aws-secret-access-key: xxx=
>   
> # kubectl apply -f ./secrets.nocommit.yaml
> ```
> 
> What do you think ?

I agree with this by the way. The secret should be a prerequisite, not a configuration parameter.

File: chart/values.yaml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+      awsAccessKeyIdKey: AWS_ACCESS_KEY_ID
+      awsSecretAccessKeyKey: AWS_SECRET_ACCESS_KEY
+    #
+    # Or provide AWS AK/SK (helm 3 values are saved as secrets)
+    accessKeyId: null
+    secretAccessKey: null
+
+image:
+  repository: nabsul/k8s-ecr-login-renew
+  tag: latest
```
Comment (nabsul) [inline chart/values.yaml:23]: I think this should be hard-coded to a specific version. To be honest, I regret ever releasing a `latest` tag :-). I don't want anyone to take a dependency on that.

File: chart/values.yaml [hunk 14]
```diff
@@ -0,0 +1,24 @@
+secretName: docker-ecr
+# target Namespace defaults to release namespace if left unspecified
+# targetNamespace: "*"
+
+ecr:
+  region: us-east-1
```
Comment (nabsul) [inline chart/values.yaml:6]: Probably best to force the user to enter a region explicitly?

File: index.yaml [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    - ecr
+    - imagePullSecrets
+    maintainers:
+    - email: james@tachibana.systems
+      name: James "ec0" Hebden
+    name: k8s-ecr-login-renew
+    sources:
+    - https://github.com/nabsul/k8s-ecr-login-renew
+    urls:
+    - https://nabsul.github.io/k8s-ecr-login-renew/k8s-ecr-login-renew-0.1.tgz
```
Comment (nabsul) [inline index.yaml:20]: Instead of checking the package into the repo like this, how about we try to follow this recommendation: https://docs.helm.sh/docs/howto/chart_releaser_action/

File: chart/templates/definitions.tpl [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+{{- default (printf "%s-cron" .Release.Name) -}}
+{{- end }}
+
+{{- define "k8s-ecr-login-renew.secret" }}
+{{- .Values.ecr.auth.existingSecret | default (printf "%s-secret" .Release.Name) -}}
+{{- end }}
+
+{{- define "k8s-ecr-login-renew.targetNamespace" }}
+{{- default .Release.Namespace .Values.targetNamespace -}}
+{{- end }}
```
Comment (nabsul) [inline chart/templates/definitions.tpl:35]: nit: add new line

File: README.md [hunk 17]
```diff
@@ -79,27 +79,72 @@ Ideally you should set up a policy that:
...
 kubectl create secret -n ns-ecr-renew-demo generic ecr-renew-cred-demo \
-  --from-literal=REGION=[AWS_REGION] \
-  --from-literal=ID=[AWS_KEY_ID] \
-  --from-literal=SECRET=[AWS_SECRET]
+  --from-literal=aws-access-key-id=[AWS_KEY_ID] \
+  --from-literal=aws-secret-access-key=[AWS_SECRET]
 ```
 
-### Required Kubernetes Service Account
+### Deploy to Kubernetes with Helm 3
```
Comment (nabsul) [inline README.md:103]: Probably best to add a note before this section that says something like "there are two ways to deploy this tool, and you only need to use one of these..."

**Review by nabsul — CHANGES_REQUESTED** [review]
This is a fantastic start! I didn't have a chance to fully review and test thing, but I have started reading the Helm docs, and based on what I've learned so far, I have a few suggestions.

I didn't have time today, but I'd be happy to make some of these changes myself. I think I can submit suggested changes as pull requests into your branch...

File: .gitignore [hunk 18]
```diff
@@ -2,3 +2,4 @@
 .idea
 k8s-ecr-login-renew.exe
 k8s-ecr-login-renew
+*~
```
[the remark this answers is not in the document]
Comment (Startouf) [reply .gitignore:5]: I think we can remove this

File: README.md [hunk 19]
```diff
@@ -25,7 +25,7 @@ Or by tag:
 
 ## Environment Variables
 
-The tool is mainly configured through environment variables. These are:
+The tool is mainly configured through environment variables (and/or secrets in kubernetes). These are:
```
[the remark this answers is not in the document]
Comment (Startouf) [reply README.md:28]: My bad, this is a leftover and can be removed/reverted

File: chart/values.yaml (hunk 14 above)
Comment (Startouf) [reply chart/values.yaml:6]: This is the default AWS region configured in most AWS APIs. 
There is little problem here leaving this default value, as using an ECR repository in a region that does not exist should trigger a meaningful error message from AWS.

But we can also leave it blank as you suggest, I have just not tested what error is given if this is left empty, I hope it's meaningful, and not just a "failed to parse nil as string" or something like this. I'm also the first to be annoyed that us-east-1 is the default region in some tools since I'm in Europe and sometimes end up creating resources in US regions I don't want because of this default.

File: index.yaml (hunk 15 above)
Comment (Startouf) [reply index.yaml:20]: I was not aware of this documentation, this would be better indeed.

**Comment by Startouf** [comment]
@devec0 can you handle nabsul's review requested changes ?

File: .gitignore (hunk 18 above)
[the remark this answers is not in the document]
Comment (devec0) [reply .gitignore:5]: It's for emacs backup files

File: chart/values.yaml (hunk 14 above)
Comment (devec0) [reply chart/values.yaml:6]: We can use the `required` template filter here to gracefully handle this. The general pattern I've seen with helm charts is that it is better to have a default value but I agree with having this as a required value as while most people will have their registry in us-east-1, many won't.

File: README.md (hunk 17 above)
Comment (devec0) [reply README.md:103]: Agreed, I've clarified this in the README

File: chart/templates/definitions.tpl (hunk 16 above)
Comment (devec0) [reply chart/templates/definitions.tpl]: done

File: index.yaml (hunk 15 above)
Comment (devec0) [reply index.yaml:20]: I've added the workflow - I _think_ the required environment variables should already be set by GitHub

File: chart/values.yaml (hunk 13 above)
Comment (devec0) [reply chart/values.yaml]: I've set the version to the latest tag, v1.7.1

File: .gitignore [hunk 20]
```diff
@@ -2,3 +2,4 @@
 .idea
```
[the remark this answers is not in the document]
Comment (devec0) [reply .gitignore:3]: I've made the .gitignore and .helmignore consistent and added comments explaining the entries

File: README.md [hunk 21]
```diff
@@ -79,27 +79,72 @@ Ideally you should set up a policy that:
...
+- (Optionally) Create a group for that user
+- Authorize the group or user to pull images from ECR. Either
+  a. Use the existing AWS policy "AmazonEC2ContainerRegistryReadOnly"
+  b. or create a policy for that group or user with the following configuration:
+    - Service: Elastic Container Registry
+    - Access Level: List & Read
+    - Resources: Select the specific ECR instance that you'll be using
   
-Once that's created, you'll want to get the access key ID and secret for the next step.
+Once that's created, you'll want to get the Access Key ID and Secret access Key (AK/SK) of the user for the next step.
```
[the remark this answers is not in the document]
Comment (devec0) [reply README.md]: done

File: README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (devec0) [reply README.md]: done

**Comment by devec0** [comment]
> > I suggest not creating the secret via values to preserve confidentiality. A note could be added to the readme explaining how to create the secret
> > ```
> > # secrets.nocommit.yaml
> > apiVersion: v1
> > kind: Secret
> > metadata:
> >   name: aws-secrets
> >   namespace: xxx
> > type: Opaque
> > data:
> >   aws-acces-key-id: xxx=
> >   aws-secret-access-key: xxx=
> >   
> > # kubectl apply -f ./secrets.nocommit.yaml
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What do you think ?
> 
> I agree with this by the way. The secret should be a prerequisite, not a configuration parameter.

Also, I wanted to add to this specifically - the values support is important when working with other tools and things like the helm-secrets plugin to do proper secrets management. I think having both is the right call if you want to be able to supply the secret as a pre-existing definition configured by some other tool (like terraform). Additionally, helm3 stores the values as secrets so there is no downside from a security perspective either.

**Comment by nabsul** [comment]
Heyo! I'm going to take this into a local branch on this repo to make it easier to continue the work.

**Comment by nabsul** [comment]
https://github.com/nabsul/k8s-ecr-login-renew/pull/38

**Comment by nabsul** [comment]
Thanks again for your contributions here! I believe I have https://github.com/nabsul/k8s-ecr-login-renew/pull/38 working the I want it to work. This PR was a great base to start from, but I decided to deviate a bit in the following ways:

- Avoid pre-hooks and just have all resource created in the main event
- Skip the one-off job creation in favor of giving the user the one-liner kubectl command needed to run the cronjob once manually
- I tried Helm's recommended GH action for publishing the repo, but the more I understand how repos work, the more I feel like it's not a great solution. The main reasons being:
  - The Helm chart for this tool should live in the same repo as the actual tool/code
  - The repo index however can accommodate multiple tools/projects (I plan to add a Helm chart for KCert), so the `index.yaml` file should NOT live in this repo. I have moved it to [`https://github](https://nabsul.github.io/helm/index.yaml)
  - I don't believe in checking tgz files into the git repo. Right now I'm manually creating the tgz file and uploading it as a release. In a later step, I'll create a GH action that does this automatically.
  - Automating the `index.yaml` file would be nice, but for now I'll manually edit it whenever there's a new release.
- I still would like to provide a raw-yaml approach for people who prefer `kubectl apply`, but I'll have that code auto-generated from Helm's `--debug` and dry-run commands. This will eliminate the need to maintain two separate ways to install the tool.

**Comment by devec0** [comment]
@nabsul - I'll close this one out in favour of #38
